lib/upipe-modules/upipe_block_to_sound.c

```diff
@@ -0,0 +1,315 @@
+{
+
+    struct upipe_block_to_sound *upipe_block_to_sound = upipe_block_to_sound_from_upipe(upipe);
+
+    /* get block size */
+    size_t block_size = 0;
+    uref_block_size(uref, &block_size);
+
+    /* drop incomplete samples */
+    if ((block_size % upipe_block_to_sound->sample_size) != 0) {
+        upipe_err(upipe, "Incomplete samples detected");
+    }
+
+    int samples = block_size / upipe_block_to_sound->sample_size;
+    block_size = samples * upipe_block_to_sound->sample_size;
+
+    /* alloc sound ubuf */
+    assert(upipe_block_to_sound->ubuf_mgr);
+    struct ubuf *ubuf_block_to_sound = ubuf_sound_alloc(upipe_block_to_sound->ubuf_mgr,
+                                                        samples);
+
+    if (unlikely(ubuf_block_to_sound == NULL)) {
+        uref_free(uref);
+        upipe_throw_fatal(upipe, UBASE_ERR_ALLOC);
+        return;
+    }
+
+
+    /* map sound ubuf for writing */
+    int32_t *w;
+    ubuf_sound_write_int32_t(ubuf_block_to_sound, 0, -1, &w, upipe_block_to_sound->planes);
+    /* map block ubuf for reading */
+    const uint8_t *r;
+    int end = -1;
+    uref_block_read(uref, 0, &end, &r);
+    /* copy block to sound */
+    memcpy(w, r, block_size);
+    /* unmap ubufs */
+    ubuf_sound_unmap(ubuf_block_to_sound, 0, -1, upipe_block_to_sound->planes);
+    ubuf_block_unmap(uref->ubuf, 0);
+    /* attach sound ubuf to uref */
+    uref_attach_ubuf(uref, ubuf_block_to_sound);
+    /* output pipe */
+
+    upipe_block_to_sound_output(upipe, uref, upump_p);
+}
+
+/** @internal @This sets the input flow definition.
+ *
+ * @param upipe description structure of the pipe
+ * @param flow_def flow definition packet
+ * @return an error code
+ */
+static int upipe_block_to_sound_set_flow_def(struct upipe *upipe,
+                                       struct uref *flow_def)
+{
+    struct upipe_block_to_sound *upipe_block_to_sound = upipe_block_to_sound_from_upipe(upipe);
+
+    if (flow_def == NULL)
+        return UBASE_ERR_INVALID;
+
+    const char *def;
+    if (unlikely(!ubase_check(uref_flow_get_def(flow_def, &def)) ||
+                (ubase_ncmp(def, "block."))))
+        return UBASE_ERR_INVALID;
+
+    flow_def = uref_dup(upipe_block_to_sound->flow_def_config);
+    if (unlikely(flow_def == NULL)) {
+        upipe_throw_fatal(upipe, UBASE_ERR_ALLOC);
+        return UBASE_ERR_ALLOC;
+    }
+
+    if(unlikely(!ubase_check(uref_flow_get_def(flow_def, &def))) ||
+                (ubase_ncmp(def, "sound.s32."))) {
+        return UBASE_ERR_INVALID;
+    }
+    uint8_t channels, planes, sample_size;
+    UBASE_RETURN(uref_flow_get_def(flow_def, &def))
+    UBASE_RETURN(uref_sound_flow_get_channels(flow_def, &channels))
+    UBASE_RETURN(uref_sound_flow_get_planes(flow_def, &planes))
+    UBASE_RETURN(uref_sound_flow_get_sample_size(flow_def, &sample_size))
+
+    upipe_block_to_sound_store_flow_def(upipe, uref_dup(flow_def));
```

Comment:
you're leaking an uref here

Reply:
Are you sure ? The test passes with valgrind --leak-check=full

Check 0632714f23ba4293bf98c288f248d6272187b4b0 the flow_def is reused below as ->flow_def_config